### ISSUE TITLE
Dino fetches the whole MAM history since the last launch, which takes a lot of time and CPU

### ISSUE THREAD

**Issue by marmistrz** [dino/dino#372]
I haven't used Dino for a long time - I'm mostly a Gajim user, but I still kept it installed.

After I've launched Dino, it kept fetching MAM messages for a quite long time, consuming 100% CPU. OTOH, Conversations fetches the MAM for a specific contact when the conversation with the specific contact is first opened. 

This may be a better strategy, besides, it may be wiser to only fetch the most recent MAM.

**Comment by ivucica** [comment]
I started Dino for the first time ever, and it began syncing MAM messages for various contacts since the first message I've ever had archived, back in 2014.

If full sync is desired, perhaps a better strategy would be to sync from the newest to the oldest? Though, I suspect what @Ape suggested (syncing back in time only when user performs scrollback) is probably a better strategy.

**Comment by stevenroose** [comment]
> I started Dino for the first time ever, and it began syncing MAM messages for various contacts since the first message I've ever had archived, back in 2014.

This is a huge problem. I'm making a new issue for this.

**Comment by licaon-kter** [comment]
Can Dino, at least, not fetch MAM for contacts that are NOT actually opened?

**Comment by Echolon** [comment]
Rel. #417 #772

**Comment by stevenroose** [comment]
> Can Dino, at least, not fetch MAM for contacts that are NOT actually opened?

That doesn't make any sense. How would you know then if a contact that was not open sent you a new message?

**Comment by licaon-kter** [comment]
@stevenroose because that's MAM and NOT new messages. Because I read them **elsewhere**.

**Comment by ivucica** [comment]
@licaon-kter Unfortunately, that doesn't make sense. Clients cannot know you've read them elsewhere until it has accessed the messages, including the marked-as-read message.

It can know that a client has received this message (as it has been purged from offline storage that may or may not be exposed by the server), but it cannot know that the message has or has not been read.

Use case:

1. My phone is online, however, it's left in my drawer.
2. A message is sent to me. I may or may not hear the notification sound.
3. I am walking to my desktop and I am starting Dino IM.
4. I expect to see the message on Dino, as I have not seen it yet -- neither on the phone or on Dino. 

The only way for Dino to show this message is to do a MAM query, to fetch the message and then to show it.

And for that matter, proposal you have made is unfortunately not solving this issue. One of the windows that Dino is fetching is actually a chatroom where I have 4+ years of logs stored server-side. I am storing it intentionally, it gets stored even when I am myself offline, but I don't want all of it replicated locally. 

To summarize:

- Dino needs to make some MAM query for offline messages
- It should be possible to restrict Dino to some reasonable period (last 30 messages + infinite scrollback? last 7 days of messages?), but it currently isn't

- If the time-or-count restriction is implemented, per this issue, it doesn't matter if closed contacts are synced (unless you have thousands of contacts in your roster)
- If the don't-sync-closed-contacts is implemented, I would not get to see the message delivered to my phone when I was offline unless I also opened chat with the contact who sent the message (and I don't know who sent me the message, so I cannot do that)
